osquery/examples/example_writable_table.cpp

```diff
@@ -0,0 +1,231 @@
+  }
+
+  /// Expands a value list returned by osquery into a RowData structure
+  bool getRowData(RowData& row_data, const std::string& value_list) const {
+    row_data = {};
+
+    rapidjson::Document document;
+    document.Parse(value_list);
+    if (document.HasParseError() || !document.IsArray()) {
+      VLOG(1) << "Invalid format";
+      return false;
+    }
+
+    if (document.Size() != 2U) {
+      VLOG(1) << "Wrong column count.";
+      return false;
+    }
+
+    if (!document[0].IsNull()) {
+      row_data.text = document[0].GetString();
+    }
+
+    if (document[0].IsNull()) {
+      row_data.integer = 0;
+    } else {
+      row_data.integer = document[1].GetInt();
+    }
+
+    return true;
+  }
+
+  /// Generates the rows for osquery
+  QueryData generate(QueryContext& context) {
+    QueryData results;
+
+    for (const auto& row_descriptor : row_map) {
+      const auto row_data = row_descriptor.second;
+
+      Row r = {};
+
+      // The rowid is optional, but we need to populate it if we want to
+      // correctly match our structures
+      r["rowid"] = std::to_string(row_data.rowid);
+      r["text"] = row_data.text;
+      r["integer"] = INTEGER(row_data.integer);
+
+      results.push_back(r);
+    }
+
+    return results;
+  }
+
+  /// Callback for DELETE queries
+  QueryData delete_(QueryContext& context, const PluginRequest& request) {
+    const auto& rowid = request.at("id");
+
+    auto internal_id_it = id_to_internal_id.find(rowid);
+    if (internal_id_it == id_to_internal_id.end()) {
+      VLOG(1) << "The rowid is not mapped to an internal rowid";
+      return {{std::make_pair("status", "failure")}};
+    }
+
+    auto internal_id = internal_id_it->second;
+
+    auto row_it = row_map.find(internal_id);
+    if (row_it == row_map.end()) {
+      VLOG(1) << "Internal error; row id -> primary key mismatch";
+      return {{std::make_pair("status", "failure")}};
+    }
+
+    row_map.erase(row_it);
+    return {{std::make_pair("status", "success")}};
+  }
+
+  /// Callback for INSERT queries
+  QueryData insert(QueryContext& context, const PluginRequest& request) {
+    // This variable contains a value for each column in the table, separated by
+    // ';'
+    const auto& value_list = request.at("value_list");
+
+    RowData row_data;
+    if (!getRowData(row_data, value_list)) {
+      return {{std::make_pair("status", "failure")}};
+    }
+
+    // If auto_rowid is false, then we get to decide which rowid to use
+    std::uint64_t new_rowid;
+    if (request.at("auto_rowid") == "false") {
+      // Do not increment the rowid generator until we accept the data
+      // we have received!
+      new_rowid = rowid_generator + 1;
+
+    } else {
+      const auto& id = request.at("id");
+      new_rowid = std::strtoull(id.data(), nullptr, 10);
+    }
+
+    row_data.rowid = new_rowid;
+
+    // Build the primary key and make sure it's unique
+    std::string internal_rowid =
```

Comment:
This should be abstracted as a function that generates the primary key of the table. Currently adding a new column to the primary key is going to require changes to places and will be error prone (even if you use all the right columns but in the wrong order).